File: Problem2.py
#Time complexity is O(n^3)
#Space complexity is O(n^2)
#Code ran suceessfully on leetcode
#No issues faced while coding
class Solution(object):
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #Base condition
        if(nums==None or len(nums)==0):
            return 0
        #Initializing all the required variables
        n=len(nums)
        dp=[[0 for i in range(n)] for i in range(n)]
        #Looping through all the values in the given nums array
        for l in range(1,n+1):
            #start index of the burstible array
            for i in range(0,n-l+1):
                #i is the starting index
                #end index of the burstible array, j is the end index
                j=i+l-1
                #Taking the maxi value as the minimim infinity
                maxi=-1*float('inf')
                #Looping the array under consideration
                for k in range(i,j+1):
                    #Taking left and right values
                    left=0
                    right=0
                    if(i!=k):
                        #If the above condition satisfies, below one will be left value
                        left=dp[i][k-1]
                    if(j!=k):
                        #If the above condition satisfies, below one will be left value
                        right=dp[k+1][j]
                    #Initialzing before and after values
                    before=1
                    after=1
                    if(i!=0):
                        #If the above condition satisfies, below one will be before value
                        before=nums[i-1]
                    if(j!=n-1):
                        #If the above condition satisfies, below one will be after value
                        after=nums[j+1]
                    #We will take the maxi mum value at each step
                    maxi=max(maxi,left+before*nums[k]*after+right)
                #FInally we will update the dp matrix with maximum value
                dp[i][j]=maxi
        #We will return value in the first row last clomun
        return dp[0][n-1]

File: problem2.cpp
/*
// Time Complexity : o(N^3)
// Space Complexity : O(N^2)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
from bottom up apporach going from the last ballon left to all the ballons

*/

#include<iostream>
#include<vector>

using namespace std;

// this one is using memorization 

class Solution {
    vector<vector<int>> dp_arr;
    int dp(int i,int j,vector<int>& nums){
        if(i>j) return 0;
        if(dp_arr[i][j]!=-1) return dp_arr[i][j];
        int maxi{INT_MIN};
        for(int idx{i};idx<=j;++idx){
            int cost = nums.at(i-1)*nums.at(idx)*nums.at(j+1) + dp(i,idx-1,nums) + dp(idx+1,j,nums);
            maxi = max(maxi,cost);
        }   
        return dp_arr[i][j] = maxi;
    }
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        dp_arr.resize(n+1,vector<int>(n+1,-1));
        nums.push_back(1);
        nums.insert(nums.begin(),1);
        return dp(1,n,nums);
    }
};

//the one with tabulation 


class Solution{
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp_arr(n+2,vector<int>(n+2,0));
        nums.push_back(1);
        nums.insert(nums.begin(),1);
        for(int i{n};i>=1;--i){
            for(int j{1};j<=n;++j)
            {
                if(i>j) continue;
                int maxi{INT_MIN};
                for(int idx{i};idx<=j;++idx){
                    int cost = nums.at(i-1)*nums.at(idx)*nums.at(j+1) + dp_arr.at(i).at(idx-1) + dp_arr.at(idx+1).at(j);
                    maxi = max(maxi,cost);
                }
                dp_arr[i][j] = maxi;
            }
        }
        return dp_arr.at(1).at(n);
    }
};